www/www.reactos.org/roslogin/admin/Mattermost.php

```diff
@@ -0,0 +1,228 @@
+<?php
+/*
+ * PROJECT:     RosLogin - A simple Self-Service and Single-Sign-On around an LDAP user directory
+ * LICENSE:     AGPL-3.0-or-later (https://spdx.org/licenses/AGPL-3.0-or-later)
+ * PURPOSE:     Mattermost communication with RosAdmin
+ * COPYRIGHT:   Copyright 2020 Stanislav Motylkov (x86corez@gmail.com)
+ */
+
+	class Mattermost
+	{
+		//
+		// MEMBER VARIABLES
+		//
+		private $_ch;
+		private $_url;
+		private $_last_error;
+
+		//
+		// PRIVATE FUNCTIONS
+		//
+
+		/**
+		 * Is this error returned by Mattermost server?
+		 */
+		private function isError($json)
+		{
+			return (isset($json['message']) && isset($json['status_code']));
+		}
+
+		/**
+		 * Generates unique to the client temporary filename
+		 */
+		private function getTempFilename()
+		{
+			$uniq = md5($_SERVER['REMOTE_ADDR'] . $_SERVER['HTTP_USER_AGENT']);
+			return sys_get_temp_dir() . '/rosadmin_mattermost_' . substr($uniq, 0, 8);
+		}
+
+		/**
+		 * Reads error message that was saved earlier
+		 * and removes it from disk
+		 */
+		private function restoreError()
+		{
+			$filename = Mattermost::getTempFilename();
+			if (!file_exists($filename))
+				return null;
+			$error = file_get_contents($filename);
+			unlink($filename);
+			return $error;
+		}
+
+		//
+		// PUBLIC FUNCTIONS
+		//
+
+		public function __construct($curl_helper, $url, $token)
+		{
+			$this->_ch = $curl_helper;
+			$this->_url = $url;
+			$this->_ch->setHeaders(array("Authorization: Bearer {$token}"));
+			$this->_last_error = null;
+		}
+
+		/**
+		 * Returns user ID
+		 */
+		public function getUserByName($name)
+		{
+			/* Mattermost usernames are lower-case */
+			$name = urlencode(strtolower($name));
```

Comment:
Mattermost usernames also have spaces converted to dashes (e.g. "Colin Finck" becomes "colin-finck").
I haven't found the original code that has all the rules (it may be closed source anyway), but I don't know of any further rule either.